pulsar-jms-tracing/src/main/java/com/datastax/oss/pulsar/jms/tracing/BrokerTracing.java

```diff
@@ -0,0 +1,553 @@
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.mledger.Entry;
+import org.apache.pulsar.broker.PulsarService;
+import org.apache.pulsar.broker.intercept.BrokerInterceptor;
+import org.apache.pulsar.broker.service.Consumer;
+import org.apache.pulsar.broker.service.Producer;
+import org.apache.pulsar.broker.service.ServerCnx;
+import org.apache.pulsar.broker.service.Subscription;
+import org.apache.pulsar.broker.service.Topic;
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.apache.pulsar.common.api.proto.BaseCommand;
+import org.apache.pulsar.common.api.proto.CommandAck;
+import org.apache.pulsar.common.api.proto.MessageIdData;
+import org.apache.pulsar.common.api.proto.MessageMetadata;
+import org.apache.pulsar.common.intercept.InterceptException;
+import org.apache.pulsar.common.naming.TopicName;
+import org.jetbrains.annotations.NotNull;
+
+@Slf4j
+public class BrokerTracing implements BrokerInterceptor {
+
+  public enum EventReasons {
+    ADMINISTRATIVE,
+    COMMANDS,
+    MESSAGE,
+    TRANSACTION,
+    SERVLET,
+  }
+
+  private static final TraceLevel defaultTraceLevel = TraceLevel.BASIC;
+
+  private final Set<EventReasons> jmsTracingEventList = new HashSet<>();
+  private TraceLevel traceLevel = defaultTraceLevel;
+  private int maxBinaryDataLength = 256;
+  private int cacheTraceLevelsDurationSec = 10;
+  private boolean traceSystemTopics = false;
+  private boolean traceSchema = false;
+  private boolean reduceLevelForNestedComponents = true;
+
+  private static Set<EventReasons> loadEnabledEvents(
+      PulsarService pulsarService, Set<EventReasons> enabledEvents) {
+    String events =
+        pulsarService.getConfiguration().getProperties().getProperty("jmsTracingEventList", "");
+    log.debug("read jmsTracingEventList: {}", events);
+
+    for (String event : events.split(",")) {
+      try {
+        enabledEvents.add(EventReasons.valueOf(event.trim().toUpperCase()));
+      } catch (IllegalArgumentException e) {
+        log.error("Invalid event: {}. Skipping", event);
+      }
+    }
+
+    return enabledEvents;
+  }
+
+  @NotNull
+  private static TraceLevel getTraceLevel(PulsarService pulsar) {
+    String level =
+        pulsar
+            .getConfiguration()
+            .getProperties()
+            .getProperty("jmsTracingLevel", defaultTraceLevel.toString());
+    try {
+      return TraceLevel.valueOf(level.trim().toUpperCase());
+    } catch (IllegalArgumentException e) {
+      log.warn("Invalid tracing level: {}. Using default: {}", level, defaultTraceLevel);
+      return defaultTraceLevel;
+    }
+  }
+
+  private final LoadingCache<Subscription, TraceLevel> traceLevelForSubscription =
```

Comment:
what about using a AsyncLoading cache ? we are going to issue network calls to get topic properties and there is an async API in PulsarAdmin

Reply:
I added async refresh.
As long as topic/sub is used, the cache will refresh values in background (returning old values until refresh is done) and it will reset write timestamp so expiration will postone as well.
I can return NONE on the first call and let cache reload value later but at the moment this does not look like it will be a frequent case, also this complicates testing etc.